models/classes/runner/OfflineQtiRunnerService.php

```diff
@@ -0,0 +1,147 @@
+        $items = [];
+
+        foreach ($this->getItemIdentifiersFromTestMap($testMap) as $itemIdentifier) {
+            $itemRef = $runnerService->getItemHref($serviceContext, $itemIdentifier);
+
+            /** @var QtiRunnerServiceContext $serviceContext */
+            $items[$itemIdentifier] = [
+                'baseUrl' => $runnerService->getItemPublicUrl($serviceContext, $itemRef),
+                'itemData' => $this->getItemData($serviceContext, $itemRef),
+                'itemState' => $runnerService->getItemState($serviceContext, $itemIdentifier),
+                'itemIdentifier' => $itemIdentifier,
+                'portableElements' => $runnerService->getItemPortableElements($serviceContext, $itemRef),
+            ];
+        }
+
+        return $items;
+    }
+
+    /**
+     * Returns the itemData, extending with the variable elements
+     *
+     * @param RunnerServiceContext $context
+     * @param string $itemRef
+     * @return array
+     * @throws \common_exception_InvalidArgumentType
+     * @throws \common_Exception
+     */
+    public function getItemData(RunnerServiceContext $context, $itemRef)
+    {
+        $this->getRunnerService()->assertQtiRunnerServiceContext($context);
+
+        $itemData = $this->getRunnerService()->getItemData($context, $itemRef);
+        $itemDataVariable = $this->getRunnerService()->getItemVariableElementsData($context, $itemRef);
+        $responses = $itemData['data']['responses'];
+
+        foreach (array_keys($responses) as $responseId) {
+            if (array_key_exists($responseId, $itemDataVariable)) {
+                $itemData['data']['responses'][$responseId] = array_merge(...[
+                    $responses[$responseId],
+                    $itemDataVariable[$responseId],
+                ]);
+            }
+        }
+
+        return $itemData;
+    }
+
+    /**
+     * Returns the item identifiers
+     *
+     * @param array $testMap
+     * @return array
+     */
+    private function getItemIdentifiersFromTestMap($testMap)
+    {
+        return $this->getSubIdentifiersRecursively($testMap, [
+            RunnerMap::MAP_ATTRIBUTE_PARTS,
+            RunnerMap::MAP_ATTRIBUTE_SECTIONS,
+            RunnerMap::MAP_ATTRIBUTE_ITEMS,
+        ]);
+    }
+
+    /**
+     * Calls itself recursively to return identifiers from nested arrays
+     *
+     * @param $array
+     * @param $identifiers
+     * @return array
+     */
+    private function getSubIdentifiersRecursively($array, $identifiers)
+    {
+        $identifier = array_shift($identifiers);
+
+        if (count($identifiers) > 0) {
+            $result = [];
+
+            foreach ($array[$identifier] as $key => $value) {
+                $result[] = $this->getSubIdentifiersRecursively(
+                    $array[$identifier][$key],
+                    $identifiers
+                );
+            }
+
+            return array_merge(...$result);
```

Comment:
I don't know if we still support it, but this won't work below php 5.6

Reply:
As far as I know we should support PHP5.6 or higher. @siwane can you please confirm or not? :) Thanks!

Reply:
I agree to use it, it is compatible with PHP5.6 and we support lower than this version.